src/test/rustdoc/intra-doc-crate/traits.rs

```diff
@@ -0,0 +1,15 @@
+// aux-build:traits.rs
+// build-aux-docs
+// ignore-tidy-line-length
+#![deny(intra_doc_link_resolution_failure)]
+
+extern crate inner;
+use inner::SomeTrait;
+
+pub struct SomeStruct;
+
+ // @has 'traits/struct.SomeStruct.html' '//a[@href="../inner/trait.SomeTrait.html"]' 'SomeTrait'
+impl SomeTrait for SomeStruct {
+    // @has 'traits/struct.SomeStruct.html' '//a[@href="../inner/trait.SomeTrait.html"]' 'SomeTrait'
```

Comment:
This check isn't specific enough. It's picking up the `SomeTrait` link in `impl SomeTrait for SomeStruct`, so the check passes but the docs include a link resolution failure.

![foo](https://user-images.githubusercontent.com/1372438/87240346-bb455900-c3e6-11ea-9e4a-d22c245bb918.png)

Reply:
Yes, you're quite right. I'm not sure why the link is broken, let me debug this some more.

Reply:
The more concerning thing IMO is that `deny(intra_doc_link_resolution_failure)` didn't trigger :(

Reply:
This is somehow related to https://github.com/rust-lang/rust/issues/56546. The `collect_intra_doc_links` pass doesn't see the documentation from the trait _at all_, it only sees the docs that were written on the trait implementation. Let me try this with the trait in the same crate, I expect that will also reproduce the bug.

Reply:
Hmm, in the same crate it works ... I'm very confused what's going on.

Reply:
Since this looks tricky I think I'll delay it to a follow-up MR unless I happen to figure it before #73566 lands. Here is the diff to fix the test case:

```rust
diff --git a/src/test/rustdoc/intra-doc-crate/auxiliary/traits.rs b/src/test/rustdoc/intra-doc-crate/auxiliary/traits.rs
index 5c3360dc5fd..c16e39d56f3 100644
--- a/src/test/rustdoc/intra-doc-crate/auxiliary/traits.rs
+++ b/src/test/rustdoc/intra-doc-crate/auxiliary/traits.rs
@@ -1,7 +1,7 @@
 #![crate_name = "inner"]
 /// this is a trait
 pub trait SomeTrait {
-    /// this is a method for [SomeTrait]
+    /// this is a method for [a trait][SomeTrait]
     fn foo();
 }
 
diff --git a/src/test/rustdoc/intra-doc-crate/traits.rs b/src/test/rustdoc/intra-doc-crate/traits.rs
index 0672c76435b..0fcd77e4019 100644
--- a/src/test/rustdoc/intra-doc-crate/traits.rs
+++ b/src/test/rustdoc/intra-doc-crate/traits.rs
@@ -10,6 +10,6 @@ pub struct SomeStruct;
 
  // @has 'traits/struct.SomeStruct.html' '//a[@href="../inner/trait.SomeTrait.html"]' 'SomeTrait'
 impl SomeTrait for SomeStruct {
-    // @has 'traits/struct.SomeStruct.html' '//a[@href="../inner/trait.SomeTrait.html"]' 'SomeTrait'
+    // @has 'traits/struct.SomeStruct.html' '//a[@href="../inner/trait.SomeTrait.html"]' 'a trait'
     fn foo() {}
 }
```